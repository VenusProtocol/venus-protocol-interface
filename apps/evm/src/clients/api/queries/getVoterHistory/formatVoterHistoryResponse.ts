import BigNumber from 'bignumber.js';

import type { VoterHistory } from 'types';
import { formatToProposal } from './formatToProposal';

import type { GetVoterHistoryResponse } from './types';

const formatVoterHistoryResponse = (
  data: GetVoterHistoryResponse,
): {
  limit: number;
  page: number;
  total: number;
  voterHistory: VoterHistory[];
} => ({
  limit: data.limit,
  page: data.page,
  total: data.total,
  voterHistory: data.result.map(d => ({
    ...formatToProposal({ ...d, accountAddress: d.votes[0].address || '' }),
    support: d.votes[0].support,
    votesMantissa: new BigNumber(d.votes[0].votesMantissa),
    reason: d.votes[0].reason,
  })),
});

export default formatVoterHistoryResponse;
