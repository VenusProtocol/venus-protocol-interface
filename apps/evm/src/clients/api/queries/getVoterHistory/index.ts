import { VError } from 'libs/errors';
import { VoterHistory } from 'types';
import { restService } from 'utilities';

import formatVoterHistoryResponse from './formatVoterHistoryResponse';
import { GetVoterHistoryResponse } from './types';

export interface GetVoterHistoryInput {
  page?: number;
  address: string;
}

export interface GetVoterHistoryOutput {
  voterHistory: VoterHistory[];
  limit: number;
  total: number;
}

const getVoterHistory = async ({
  page = 0,
  address,
}: GetVoterHistoryInput): Promise<GetVoterHistoryOutput> => {
  const response = await restService<GetVoterHistoryResponse>({
    endpoint: `/governance/voters/${address}/history`,
    method: 'GET',
    params: {
      limit: 6,
      page,
    },
  });

  const payload = response.data;

  // @todo Add specific api error handling
  if (payload && 'error' in payload) {
    throw new VError({
      type: 'unexpected',
      code: 'somethingWentWrong',
      data: { message: payload.error },
    });
  }

  if (!payload) {
    throw new VError({ type: 'unexpected', code: 'somethingWentWrongRetrievingVoterHistory' });
  }

  return formatVoterHistoryResponse(payload);
};

export default getVoterHistory;
