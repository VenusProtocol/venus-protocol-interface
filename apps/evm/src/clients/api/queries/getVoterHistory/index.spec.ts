import Vi from 'vitest';

import voterHistoryResponse from '__mocks__/api/voterHistory.json';
import fakeAddress from '__mocks__/models/address';

import { restService } from 'utilities';

import getVoterHistory from '.';

vi.mock('utilities/restService');

describe('api/queries/getVoterHistory', () => {
  test('returns formatted voter history', async () => {
    (restService as Vi.Mock).mockImplementationOnce(async () => ({
      status: 200,
      data: voterHistoryResponse,
    }));

    const payload = await getVoterHistory({
      page: 2,
      address: fakeAddress,
    });

    expect(restService).toBeCalledWith({
      endpoint: `/governance/voters/${fakeAddress}/history`,
      method: 'GET',
      params: {
        limit: 6,
        page: 2,
      },
    });

    expect(payload).toMatchSnapshot();
  });

  test('Gets called with correct default arguments', async () => {
    (restService as Vi.Mock).mockImplementationOnce(async () => ({
      status: 200,
      data: voterHistoryResponse,
    }));

    const payload = await getVoterHistory({ address: fakeAddress });

    expect(payload.voterHistory).toHaveLength(6);

    expect(restService).toBeCalledWith({
      endpoint: `/governance/voters/${fakeAddress}/history`,
      method: 'GET',
      params: {
        limit: 6,
        page: 0,
      },
    });

    expect(payload).toMatchSnapshot();
  });
});
