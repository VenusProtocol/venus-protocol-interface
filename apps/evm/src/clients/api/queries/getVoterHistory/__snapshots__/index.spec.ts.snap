// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`api/queries/getVoterHistory > Gets called with correct default arguments 1`] = `
{
  "limit": 6,
  "page": 0,
  "total": 109,
  "voterHistory": [
    {
      "abstainVotes": [],
      "abstainedVotesMantissa": "0",
      "againstVotes": [],
      "againstVotesMantissa": "0",
      "cancelDate": undefined,
      "cancelTxHash": undefined,
      "createdDate": 2023-09-14T15:06:17.000Z,
      "createdTxHash": "0x7f1921dbaf457d146e151449b97287d5f6037081470d18dad0573749f0837143",
      "description": {
        "abstainDescription": "No opinion on the matter",
        "againstDescription": "Defer configuration of Liquidity Mining",
        "description": "**Summary**

If passed, this VIP will set the last block with rewards for the BNBx market in the Liquid Staked BNB pool, and the HAY market in the Stablecoins pool

**Description**

This VIP sets the last block with rewards for the BNBx market in the DeFi pool, and the HAY market in the Stablecoins pool. Rewards were enabled at block 32871078 in the [VIP-280](https://testnet-cf.venus.io/#/governance/proposal/280).

In [vBNBx_LiquidStakedBNB](https://testnet-cf.venus.io/#/isolated-pools/pool/0x596B11acAACF03217287939f88d63b51d3771704/market/0x644A149853E5507AdF3e682218b8AC86cdD62951), the last block with rewards will be 33735078. That means around 30 days (864,000 blocks) after being enabled.

In [vHAY_Stablecoins](https://testnet-cf.venus.io/#/isolated-pools/pool/0x10b57706AD2345e590c2eA4DC02faef0d9f5b08B/market/0x170d3b2da05cc2124334240fb34ad1359e34c562), the last block with rewards will be 33677478. That means around 28 days (806,400 blocks) after being enabled.

Simulation of the VIP: [https://github.com/VenusProtocol/vips/pull/70](https://github.com/VenusProtocol/vips/pull/70)",
        "forDescription": "Process to configure Liquidity Mining",
        "title": "VIP-169: Isolated Lending Pools: Configure Liquidity Mining",
        "version": "v2",
      },
      "endBlock": 33336002,
      "endDate": 2023-09-14T15:21:17.000Z,
      "executedDate": 2023-09-14T15:39:44.000Z,
      "executedTxHash": "0xac2d0b0876517e098d16aa4546b66a0f92391fb025a76fb25e8f3989e8be1dd1",
      "executionEtaDate": 2023-09-14T15:32:50.000Z,
      "forVotes": [
        {
          "address": "0x2ce1d0ffd7e869d9df33e28552b12ddded326706",
          "blockNumber": 33335869,
          "blockTimestamp": 1970-01-20T14:45:04.478Z,
          "createdAt": "2023-09-14T15:15:17.000Z",
          "id": "ad7a1255-d4ae-4c7b-bfaf-25ab12dd60fc",
          "proposalId": 290,
          "reason": "",
          "support": 1,
          "updatedAt": "2023-09-14T15:15:17.000Z",
          "votesMantissa": "6.05461e+23",
        },
      ],
      "forVotesMantissa": "6.05461e+23",
      "proposalActions": [],
      "proposalId": 290,
      "proposalType": 0,
      "proposerAddress": "0x2ce1d0ffd7e869d9df33e28552b12ddded326706",
      "queuedDate": 2023-09-14T15:22:50.000Z,
      "queuedTxHash": "0x1d8c233bd9f7074f9355fc91e092a6f2ed80e6ddde10a5eec1ba25eadff21651",
      "reason": "",
      "startDate": 2023-09-14T15:13:47.000Z,
      "state": 7,
      "support": 1,
      "totalVotesMantissa": "6.05461e+23",
      "userVoteSupport": 1,
      "votesMantissa": "6.05461e+23",
    },
    {
      "abstainVotes": [],
      "abstainedVotesMantissa": "0",
      "againstVotes": [],
      "againstVotesMantissa": "0",
      "cancelDate": undefined,
      "cancelTxHash": undefined,
      "createdDate": 2023-09-12T16:18:23.000Z,
      "createdTxHash": "0x5d25e9ba6d08af1c2569a3e116cbb697cebf61c2bfaaf8461a3f4041b0a2ebdd",
      "description": {
        "abstainDescription": "idk",
        "againstDescription": "n",
        "description": "If passed, this VIP will set the ProtocolShareReserve and Shortfall contracts to be used on each market of the Isolated Pools. It also configures the RiskFund contract, that will provide to the Shortfall contract the funds to be auctioned",
        "forDescription": "y",
        "title": "Risk fund foundation",
        "version": "v2",
      },
      "endBlock": 33279845,
      "endDate": 2023-09-12T16:33:23.000Z,
      "executedDate": 2023-09-12T16:47:38.000Z,
      "executedTxHash": "0x5b4f35529ccabfedeac5db1eea9385ab63a804c14f68c5e99a47ba96b6281f76",
      "executionEtaDate": 2023-09-12T16:46:41.000Z,
      "forVotes": [
        {
          "address": "0x2ce1d0ffd7e869d9df33e28552b12ddded326706",
          "blockNumber": 33279741,
          "blockTimestamp": 1970-01-20T14:42:16.091Z,
          "createdAt": "2023-09-12T20:17:48.000Z",
          "id": "8fcab0b4-cd4c-4a38-916e-d7e097f4c7b0",
          "proposalId": 288,
          "reason": "",
          "support": 1,
          "updatedAt": "2023-09-12T20:17:48.000Z",
          "votesMantissa": "6.05461e+23",
        },
      ],
      "forVotesMantissa": "6.05461e+23",
      "proposalActions": [],
      "proposalId": 288,
      "proposalType": 0,
      "proposerAddress": "0x2ce1d0ffd7e869d9df33e28552b12ddded326706",
      "queuedDate": 2023-09-12T16:36:41.000Z,
      "queuedTxHash": "0x32ac482bf430b49cb15e5f7860a4d75aeb23a171977034d52ade034e8c9d9bd3",
      "reason": "",
      "startDate": 2023-09-12T16:25:53.000Z,
      "state": 7,
      "support": 1,
      "totalVotesMantissa": "6.05461e+23",
      "userVoteSupport": 1,
      "votesMantissa": "6.05461e+23",
    },
    {
      "abstainVotes": [],
      "abstainedVotesMantissa": "0",
      "againstVotes": [],
      "againstVotesMantissa": "0",
      "cancelDate": undefined,
      "cancelTxHash": undefined,
      "createdDate": 2023-09-12T15:58:56.000Z,
      "createdTxHash": "0x3c0a3d59b6ca7c3413abd28fb1a388140a0838a9ef18c7af086815d00f38a764",
      "description": {
        "abstainDescription": "idk",
        "againstDescription": "n",
        "description": ".",
        "forDescription": "y",
        "title": "Fix XVS address in Comptroller",
        "version": "v2",
      },
      "endBlock": 33279456,
      "endDate": 2023-09-12T16:13:56.000Z,
      "executedDate": 2023-09-12T16:28:35.000Z,
      "executedTxHash": "0xc15918d42abe6dba39b1872784671e72769135bc32811012f4712e45ceacc0bf",
      "executionEtaDate": 2023-09-12T16:25:26.000Z,
      "forVotes": [
        {
          "address": "0x2ce1d0ffd7e869d9df33e28552b12ddded326706",
          "blockNumber": 33279324,
          "blockTimestamp": 1970-01-20T14:42:14.840Z,
          "createdAt": "2023-09-12T20:12:20.000Z",
          "id": "118d28f8-fc0a-4a46-8af5-e43f944c8fd1",
          "proposalId": 287,
          "reason": "",
          "support": 1,
          "updatedAt": "2023-09-12T20:12:20.000Z",
          "votesMantissa": "6.05461e+23",
        },
      ],
      "forVotesMantissa": "6.05461e+23",
      "proposalActions": [],
      "proposalId": 287,
      "proposalType": 0,
      "proposerAddress": "0x2ce1d0ffd7e869d9df33e28552b12ddded326706",
      "queuedDate": 2023-09-12T16:15:26.000Z,
      "queuedTxHash": "0x32b68a97e6a7431ec1b8c5834d27e1bf40ad5b0f1b7014d5ce7fb067b278bd87",
      "reason": "",
      "startDate": 2023-09-12T16:06:26.000Z,
      "state": 7,
      "support": 1,
      "totalVotesMantissa": "6.05461e+23",
      "userVoteSupport": 1,
      "votesMantissa": "6.05461e+23",
    },
    {
      "abstainVotes": [],
      "abstainedVotesMantissa": "0",
      "againstVotes": [],
      "againstVotesMantissa": "0",
      "cancelDate": undefined,
      "cancelTxHash": undefined,
      "createdDate": 2023-09-11T09:43:47.000Z,
      "createdTxHash": "0x227ac268325bb68450ac5cabdd09d9e2f6e40a0946c5a0d256c0ab0fdf22ae6b",
      "description": {
        "abstainDescription": "3",
        "againstDescription": "2",
        "description": ".",
        "forDescription": "1",
        "title": "VIP-168 Forced liquidations",
        "version": "v2",
      },
      "endBlock": 33243153,
      "endDate": 2023-09-11T09:58:47.000Z,
      "executedDate": 2023-09-11T10:34:59.000Z,
      "executedTxHash": "0x95014a018718ec888a696259d470715799249071c39e8b1d8c76da407e2de72a",
      "executionEtaDate": 2023-09-11T10:28:11.000Z,
      "forVotes": [
        {
          "address": "0x2ce1d0ffd7e869d9df33e28552b12ddded326706",
          "blockNumber": 33243051,
          "blockTimestamp": 1970-01-20T14:40:26.021Z,
          "createdAt": "2023-09-11T09:54:47.000Z",
          "id": "a561d4be-1a07-4af4-8929-d8f331766bf8",
          "proposalId": 284,
          "reason": "",
          "support": 1,
          "updatedAt": "2023-09-11T09:54:47.000Z",
          "votesMantissa": "6.05461e+23",
        },
      ],
      "forVotesMantissa": "6.05461e+23",
      "proposalActions": [],
      "proposalId": 284,
      "proposalType": 0,
      "proposerAddress": "0x2ce1d0ffd7e869d9df33e28552b12ddded326706",
      "queuedDate": 2023-09-11T10:18:11.000Z,
      "queuedTxHash": "0x4a4f3d03765ed3c1d27e9338ca57702224771a8590852a8dbf3ef590ee880271",
      "reason": "",
      "startDate": 2023-09-11T09:51:17.000Z,
      "state": 7,
      "support": 1,
      "totalVotesMantissa": "6.05461e+23",
      "userVoteSupport": 1,
      "votesMantissa": "6.05461e+23",
    },
    {
      "abstainVotes": [],
      "abstainedVotesMantissa": "0",
      "againstVotes": [],
      "againstVotesMantissa": "0",
      "cancelDate": undefined,
      "cancelTxHash": undefined,
      "createdDate": 2023-08-31T09:35:13.000Z,
      "createdTxHash": "0x540060cd0a1115422388ddc4ab7f0d938d99c65a9d638732661a28b6c2ab89cc",
      "description": {
        "abstainDescription": "I am indifferent to whether Venus Protocol proceeds with this recommendation or not",
        "againstDescription": "I do not think that Venus Protocol should proceed with this recommendation",
        "description": "
    Configure Direct Prices for LUNA, CAN and UST
    ",
        "forDescription": "I agree that Venus Protocol should proceed with this recommendation",
        "title": "VIP Resilient Oracle: Luna and UST",
        "version": "v2",
      },
      "endBlock": 32926185,
      "endDate": 2023-08-31T09:50:13.000Z,
      "executedDate": 2023-08-31T10:55:01.000Z,
      "executedTxHash": "0x684ed7132eae0ef40ac3fa13f5e66e05ce4b80cb2bc42806f423b35a4357bbfa",
      "executionEtaDate": 2023-08-31T10:52:34.000Z,
      "forVotes": [
        {
          "address": "0x2ce1d0ffd7e869d9df33e28552b12ddded326706",
          "blockNumber": 32926091,
          "blockTimestamp": 1970-01-20T14:24:35.131Z,
          "createdAt": "2023-08-31T09:45:54.000Z",
          "id": "e86b3675-8650-4ed0-af36-33b1813d7baa",
          "proposalId": 281,
          "reason": "",
          "support": 1,
          "updatedAt": "2023-08-31T09:45:54.000Z",
          "votesMantissa": "6.05461e+23",
        },
      ],
      "forVotesMantissa": "6.05461e+23",
      "proposalActions": [],
      "proposalId": 281,
      "proposalType": 0,
      "proposerAddress": "0x2ce1d0ffd7e869d9df33e28552b12ddded326706",
      "queuedDate": 2023-08-31T10:42:34.000Z,
      "queuedTxHash": "0xd01b12dda1c8fb28c37515341aac4e72e559b50e5c707d9c3e18690be558ef9f",
      "reason": "",
      "startDate": 2023-08-31T09:42:43.000Z,
      "state": 8,
      "support": 1,
      "totalVotesMantissa": "6.05461e+23",
      "userVoteSupport": 1,
      "votesMantissa": "6.05461e+23",
    },
    {
      "abstainVotes": [],
      "abstainedVotesMantissa": "0",
      "againstVotes": [],
      "againstVotesMantissa": "0",
      "cancelDate": undefined,
      "cancelTxHash": undefined,
      "createdDate": 2023-08-29T11:26:16.000Z,
      "createdTxHash": "0x341808efaa4388db1be787df9c149d6d6b5755e7cd1c5c2b96f40fe52396900e",
      "description": {
        "abstainDescription": "I am indifferent to whether Venus Protocol proceeds or not",
        "againstDescription": "I do not think that Venus Protocol should proceed with this proposal",
        "description": "Uppgrade comptroller to support rewards distribution for duplicate reward tokens",
        "forDescription": "I agree that Venus Protocol should proceed with this proposal",
        "title": "VIP-163 HAY and SD Rewards",
        "version": "v2",
      },
      "endBlock": 32870813,
      "endDate": 2023-08-29T11:41:37.000Z,
      "executedDate": 2023-08-29T11:54:52.000Z,
      "executedTxHash": "0xb06be6ed9813e4e4c675702bd74bb0f75c897c25365d9eea3387520c9ff39257",
      "executionEtaDate": 2023-08-29T11:52:16.000Z,
      "forVotes": [
        {
          "address": "0x2ce1d0ffd7e869d9df33e28552b12ddded326706",
          "blockNumber": 32870680,
          "blockTimestamp": 1970-01-20T14:21:48.898Z,
          "createdAt": "2023-08-30T01:21:07.000Z",
          "id": "191c17ae-da5f-4060-b5f6-4d9630081873",
          "proposalId": 280,
          "reason": "",
          "support": 1,
          "updatedAt": "2023-08-30T01:21:07.000Z",
          "votesMantissa": "6.05461e+23",
        },
      ],
      "forVotesMantissa": "6.05461e+23",
      "proposalActions": [],
      "proposalId": 280,
      "proposalType": 0,
      "proposerAddress": "0x2ce1d0ffd7e869d9df33e28552b12ddded326706",
      "queuedDate": 2023-08-29T11:42:16.000Z,
      "queuedTxHash": "0xb5e75d6211a9fdd794316b6369dd0052e9dfc3280dca6dbb0857ccf866cafc6e",
      "reason": "",
      "startDate": 2023-08-29T11:34:07.000Z,
      "state": 8,
      "support": 1,
      "totalVotesMantissa": "6.05461e+23",
      "userVoteSupport": 1,
      "votesMantissa": "6.05461e+23",
    },
  ],
}
`;

exports[`api/queries/getVoterHistory > returns formatted voter history 1`] = `
{
  "limit": 6,
  "page": 0,
  "total": 109,
  "voterHistory": [
    {
      "abstainVotes": [],
      "abstainedVotesMantissa": "0",
      "againstVotes": [],
      "againstVotesMantissa": "0",
      "cancelDate": undefined,
      "cancelTxHash": undefined,
      "createdDate": 2023-09-14T15:06:17.000Z,
      "createdTxHash": "0x7f1921dbaf457d146e151449b97287d5f6037081470d18dad0573749f0837143",
      "description": {
        "abstainDescription": "No opinion on the matter",
        "againstDescription": "Defer configuration of Liquidity Mining",
        "description": "**Summary**

If passed, this VIP will set the last block with rewards for the BNBx market in the Liquid Staked BNB pool, and the HAY market in the Stablecoins pool

**Description**

This VIP sets the last block with rewards for the BNBx market in the DeFi pool, and the HAY market in the Stablecoins pool. Rewards were enabled at block 32871078 in the [VIP-280](https://testnet-cf.venus.io/#/governance/proposal/280).

In [vBNBx_LiquidStakedBNB](https://testnet-cf.venus.io/#/isolated-pools/pool/0x596B11acAACF03217287939f88d63b51d3771704/market/0x644A149853E5507AdF3e682218b8AC86cdD62951), the last block with rewards will be 33735078. That means around 30 days (864,000 blocks) after being enabled.

In [vHAY_Stablecoins](https://testnet-cf.venus.io/#/isolated-pools/pool/0x10b57706AD2345e590c2eA4DC02faef0d9f5b08B/market/0x170d3b2da05cc2124334240fb34ad1359e34c562), the last block with rewards will be 33677478. That means around 28 days (806,400 blocks) after being enabled.

Simulation of the VIP: [https://github.com/VenusProtocol/vips/pull/70](https://github.com/VenusProtocol/vips/pull/70)",
        "forDescription": "Process to configure Liquidity Mining",
        "title": "VIP-169: Isolated Lending Pools: Configure Liquidity Mining",
        "version": "v2",
      },
      "endBlock": 33336002,
      "endDate": 2023-09-14T15:21:17.000Z,
      "executedDate": 2023-09-14T15:39:44.000Z,
      "executedTxHash": "0xac2d0b0876517e098d16aa4546b66a0f92391fb025a76fb25e8f3989e8be1dd1",
      "executionEtaDate": 2023-09-14T15:32:50.000Z,
      "forVotes": [
        {
          "address": "0x2ce1d0ffd7e869d9df33e28552b12ddded326706",
          "blockNumber": 33335869,
          "blockTimestamp": 1970-01-20T14:45:04.478Z,
          "createdAt": "2023-09-14T15:15:17.000Z",
          "id": "ad7a1255-d4ae-4c7b-bfaf-25ab12dd60fc",
          "proposalId": 290,
          "reason": "",
          "support": 1,
          "updatedAt": "2023-09-14T15:15:17.000Z",
          "votesMantissa": "6.05461e+23",
        },
      ],
      "forVotesMantissa": "6.05461e+23",
      "proposalActions": [],
      "proposalId": 290,
      "proposalType": 0,
      "proposerAddress": "0x2ce1d0ffd7e869d9df33e28552b12ddded326706",
      "queuedDate": 2023-09-14T15:22:50.000Z,
      "queuedTxHash": "0x1d8c233bd9f7074f9355fc91e092a6f2ed80e6ddde10a5eec1ba25eadff21651",
      "reason": "",
      "startDate": 2023-09-14T15:13:47.000Z,
      "state": 7,
      "support": 1,
      "totalVotesMantissa": "6.05461e+23",
      "userVoteSupport": 1,
      "votesMantissa": "6.05461e+23",
    },
    {
      "abstainVotes": [],
      "abstainedVotesMantissa": "0",
      "againstVotes": [],
      "againstVotesMantissa": "0",
      "cancelDate": undefined,
      "cancelTxHash": undefined,
      "createdDate": 2023-09-12T16:18:23.000Z,
      "createdTxHash": "0x5d25e9ba6d08af1c2569a3e116cbb697cebf61c2bfaaf8461a3f4041b0a2ebdd",
      "description": {
        "abstainDescription": "idk",
        "againstDescription": "n",
        "description": "If passed, this VIP will set the ProtocolShareReserve and Shortfall contracts to be used on each market of the Isolated Pools. It also configures the RiskFund contract, that will provide to the Shortfall contract the funds to be auctioned",
        "forDescription": "y",
        "title": "Risk fund foundation",
        "version": "v2",
      },
      "endBlock": 33279845,
      "endDate": 2023-09-12T16:33:23.000Z,
      "executedDate": 2023-09-12T16:47:38.000Z,
      "executedTxHash": "0x5b4f35529ccabfedeac5db1eea9385ab63a804c14f68c5e99a47ba96b6281f76",
      "executionEtaDate": 2023-09-12T16:46:41.000Z,
      "forVotes": [
        {
          "address": "0x2ce1d0ffd7e869d9df33e28552b12ddded326706",
          "blockNumber": 33279741,
          "blockTimestamp": 1970-01-20T14:42:16.091Z,
          "createdAt": "2023-09-12T20:17:48.000Z",
          "id": "8fcab0b4-cd4c-4a38-916e-d7e097f4c7b0",
          "proposalId": 288,
          "reason": "",
          "support": 1,
          "updatedAt": "2023-09-12T20:17:48.000Z",
          "votesMantissa": "6.05461e+23",
        },
      ],
      "forVotesMantissa": "6.05461e+23",
      "proposalActions": [],
      "proposalId": 288,
      "proposalType": 0,
      "proposerAddress": "0x2ce1d0ffd7e869d9df33e28552b12ddded326706",
      "queuedDate": 2023-09-12T16:36:41.000Z,
      "queuedTxHash": "0x32ac482bf430b49cb15e5f7860a4d75aeb23a171977034d52ade034e8c9d9bd3",
      "reason": "",
      "startDate": 2023-09-12T16:25:53.000Z,
      "state": 7,
      "support": 1,
      "totalVotesMantissa": "6.05461e+23",
      "userVoteSupport": 1,
      "votesMantissa": "6.05461e+23",
    },
    {
      "abstainVotes": [],
      "abstainedVotesMantissa": "0",
      "againstVotes": [],
      "againstVotesMantissa": "0",
      "cancelDate": undefined,
      "cancelTxHash": undefined,
      "createdDate": 2023-09-12T15:58:56.000Z,
      "createdTxHash": "0x3c0a3d59b6ca7c3413abd28fb1a388140a0838a9ef18c7af086815d00f38a764",
      "description": {
        "abstainDescription": "idk",
        "againstDescription": "n",
        "description": ".",
        "forDescription": "y",
        "title": "Fix XVS address in Comptroller",
        "version": "v2",
      },
      "endBlock": 33279456,
      "endDate": 2023-09-12T16:13:56.000Z,
      "executedDate": 2023-09-12T16:28:35.000Z,
      "executedTxHash": "0xc15918d42abe6dba39b1872784671e72769135bc32811012f4712e45ceacc0bf",
      "executionEtaDate": 2023-09-12T16:25:26.000Z,
      "forVotes": [
        {
          "address": "0x2ce1d0ffd7e869d9df33e28552b12ddded326706",
          "blockNumber": 33279324,
          "blockTimestamp": 1970-01-20T14:42:14.840Z,
          "createdAt": "2023-09-12T20:12:20.000Z",
          "id": "118d28f8-fc0a-4a46-8af5-e43f944c8fd1",
          "proposalId": 287,
          "reason": "",
          "support": 1,
          "updatedAt": "2023-09-12T20:12:20.000Z",
          "votesMantissa": "6.05461e+23",
        },
      ],
      "forVotesMantissa": "6.05461e+23",
      "proposalActions": [],
      "proposalId": 287,
      "proposalType": 0,
      "proposerAddress": "0x2ce1d0ffd7e869d9df33e28552b12ddded326706",
      "queuedDate": 2023-09-12T16:15:26.000Z,
      "queuedTxHash": "0x32b68a97e6a7431ec1b8c5834d27e1bf40ad5b0f1b7014d5ce7fb067b278bd87",
      "reason": "",
      "startDate": 2023-09-12T16:06:26.000Z,
      "state": 7,
      "support": 1,
      "totalVotesMantissa": "6.05461e+23",
      "userVoteSupport": 1,
      "votesMantissa": "6.05461e+23",
    },
    {
      "abstainVotes": [],
      "abstainedVotesMantissa": "0",
      "againstVotes": [],
      "againstVotesMantissa": "0",
      "cancelDate": undefined,
      "cancelTxHash": undefined,
      "createdDate": 2023-09-11T09:43:47.000Z,
      "createdTxHash": "0x227ac268325bb68450ac5cabdd09d9e2f6e40a0946c5a0d256c0ab0fdf22ae6b",
      "description": {
        "abstainDescription": "3",
        "againstDescription": "2",
        "description": ".",
        "forDescription": "1",
        "title": "VIP-168 Forced liquidations",
        "version": "v2",
      },
      "endBlock": 33243153,
      "endDate": 2023-09-11T09:58:47.000Z,
      "executedDate": 2023-09-11T10:34:59.000Z,
      "executedTxHash": "0x95014a018718ec888a696259d470715799249071c39e8b1d8c76da407e2de72a",
      "executionEtaDate": 2023-09-11T10:28:11.000Z,
      "forVotes": [
        {
          "address": "0x2ce1d0ffd7e869d9df33e28552b12ddded326706",
          "blockNumber": 33243051,
          "blockTimestamp": 1970-01-20T14:40:26.021Z,
          "createdAt": "2023-09-11T09:54:47.000Z",
          "id": "a561d4be-1a07-4af4-8929-d8f331766bf8",
          "proposalId": 284,
          "reason": "",
          "support": 1,
          "updatedAt": "2023-09-11T09:54:47.000Z",
          "votesMantissa": "6.05461e+23",
        },
      ],
      "forVotesMantissa": "6.05461e+23",
      "proposalActions": [],
      "proposalId": 284,
      "proposalType": 0,
      "proposerAddress": "0x2ce1d0ffd7e869d9df33e28552b12ddded326706",
      "queuedDate": 2023-09-11T10:18:11.000Z,
      "queuedTxHash": "0x4a4f3d03765ed3c1d27e9338ca57702224771a8590852a8dbf3ef590ee880271",
      "reason": "",
      "startDate": 2023-09-11T09:51:17.000Z,
      "state": 7,
      "support": 1,
      "totalVotesMantissa": "6.05461e+23",
      "userVoteSupport": 1,
      "votesMantissa": "6.05461e+23",
    },
    {
      "abstainVotes": [],
      "abstainedVotesMantissa": "0",
      "againstVotes": [],
      "againstVotesMantissa": "0",
      "cancelDate": undefined,
      "cancelTxHash": undefined,
      "createdDate": 2023-08-31T09:35:13.000Z,
      "createdTxHash": "0x540060cd0a1115422388ddc4ab7f0d938d99c65a9d638732661a28b6c2ab89cc",
      "description": {
        "abstainDescription": "I am indifferent to whether Venus Protocol proceeds with this recommendation or not",
        "againstDescription": "I do not think that Venus Protocol should proceed with this recommendation",
        "description": "
    Configure Direct Prices for LUNA, CAN and UST
    ",
        "forDescription": "I agree that Venus Protocol should proceed with this recommendation",
        "title": "VIP Resilient Oracle: Luna and UST",
        "version": "v2",
      },
      "endBlock": 32926185,
      "endDate": 2023-08-31T09:50:13.000Z,
      "executedDate": 2023-08-31T10:55:01.000Z,
      "executedTxHash": "0x684ed7132eae0ef40ac3fa13f5e66e05ce4b80cb2bc42806f423b35a4357bbfa",
      "executionEtaDate": 2023-08-31T10:52:34.000Z,
      "forVotes": [
        {
          "address": "0x2ce1d0ffd7e869d9df33e28552b12ddded326706",
          "blockNumber": 32926091,
          "blockTimestamp": 1970-01-20T14:24:35.131Z,
          "createdAt": "2023-08-31T09:45:54.000Z",
          "id": "e86b3675-8650-4ed0-af36-33b1813d7baa",
          "proposalId": 281,
          "reason": "",
          "support": 1,
          "updatedAt": "2023-08-31T09:45:54.000Z",
          "votesMantissa": "6.05461e+23",
        },
      ],
      "forVotesMantissa": "6.05461e+23",
      "proposalActions": [],
      "proposalId": 281,
      "proposalType": 0,
      "proposerAddress": "0x2ce1d0ffd7e869d9df33e28552b12ddded326706",
      "queuedDate": 2023-08-31T10:42:34.000Z,
      "queuedTxHash": "0xd01b12dda1c8fb28c37515341aac4e72e559b50e5c707d9c3e18690be558ef9f",
      "reason": "",
      "startDate": 2023-08-31T09:42:43.000Z,
      "state": 8,
      "support": 1,
      "totalVotesMantissa": "6.05461e+23",
      "userVoteSupport": 1,
      "votesMantissa": "6.05461e+23",
    },
    {
      "abstainVotes": [],
      "abstainedVotesMantissa": "0",
      "againstVotes": [],
      "againstVotesMantissa": "0",
      "cancelDate": undefined,
      "cancelTxHash": undefined,
      "createdDate": 2023-08-29T11:26:16.000Z,
      "createdTxHash": "0x341808efaa4388db1be787df9c149d6d6b5755e7cd1c5c2b96f40fe52396900e",
      "description": {
        "abstainDescription": "I am indifferent to whether Venus Protocol proceeds or not",
        "againstDescription": "I do not think that Venus Protocol should proceed with this proposal",
        "description": "Uppgrade comptroller to support rewards distribution for duplicate reward tokens",
        "forDescription": "I agree that Venus Protocol should proceed with this proposal",
        "title": "VIP-163 HAY and SD Rewards",
        "version": "v2",
      },
      "endBlock": 32870813,
      "endDate": 2023-08-29T11:41:37.000Z,
      "executedDate": 2023-08-29T11:54:52.000Z,
      "executedTxHash": "0xb06be6ed9813e4e4c675702bd74bb0f75c897c25365d9eea3387520c9ff39257",
      "executionEtaDate": 2023-08-29T11:52:16.000Z,
      "forVotes": [
        {
          "address": "0x2ce1d0ffd7e869d9df33e28552b12ddded326706",
          "blockNumber": 32870680,
          "blockTimestamp": 1970-01-20T14:21:48.898Z,
          "createdAt": "2023-08-30T01:21:07.000Z",
          "id": "191c17ae-da5f-4060-b5f6-4d9630081873",
          "proposalId": 280,
          "reason": "",
          "support": 1,
          "updatedAt": "2023-08-30T01:21:07.000Z",
          "votesMantissa": "6.05461e+23",
        },
      ],
      "forVotesMantissa": "6.05461e+23",
      "proposalActions": [],
      "proposalId": 280,
      "proposalType": 0,
      "proposerAddress": "0x2ce1d0ffd7e869d9df33e28552b12ddded326706",
      "queuedDate": 2023-08-29T11:42:16.000Z,
      "queuedTxHash": "0xb5e75d6211a9fdd794316b6369dd0052e9dfc3280dca6dbb0857ccf866cafc6e",
      "reason": "",
      "startDate": 2023-08-29T11:34:07.000Z,
      "state": 8,
      "support": 1,
      "totalVotesMantissa": "6.05461e+23",
      "userVoteSupport": 1,
      "votesMantissa": "6.05461e+23",
    },
  ],
}
`;
